include/vcpkg/base/cache.h

```diff
@@ -49,4 +50,32 @@ namespace vcpkg
     private:
         mutable std::map<Key, Value, Compare> m_cache;
     };
+
+    template<class Key, class Value, class Compare = std::less<>>
+    struct AtomicCache
+    {
+        template<class KeyIsh,
+                 class F,
+                 std::enable_if_t<std::is_constructible<Key, const KeyIsh&>::value &&
+                                      detail::is_callable<Compare&, const Key&, const KeyIsh&>::value,
+                                  int> = 0>
+        const Value& get_lazy(const KeyIsh& k, F&& f) const
+        {
+            auto it = m_cache.lower_bound(k);
+            // lower_bound returns the first iterator such that it->first is greater than or equal to than k, so k must
+            // be less than or equal to it->first. If k is not less than it->first, then it must be equal so we have a
+            // cache hit.
+            if (it != m_cache.end() && !(m_cache.key_comp()(k, it->first)))
+            {
+                return it->second;
+            }
+
+            std::lock_guard lock(m_mtx);
+            return m_cache.emplace_hint(it, k, static_cast<F&&>(f)())->second;
+        }
+
+    private:
+        mutable std::map<Key, Value, Compare> m_cache;
+        mutable std::mutex m_mtx;
```

Comment:
I can't add a mutex to the normal cache because this apparently needs to be copied somewhere.